instrumentation/aws-sdk/aws-sdk-1.11/src/main/java/io/opentelemetry/instrumentation/auto/awssdk/v1_11/AwsSdkClientTracer.java

```diff
@@ -26,20 +26,35 @@
 import io.grpc.Context;
 import io.opentelemetry.context.Scope;
 import io.opentelemetry.context.propagation.HttpTextFormat.Setter;
+import io.opentelemetry.instrumentation.api.Function;
+import io.opentelemetry.instrumentation.api.Functions;
+import io.opentelemetry.instrumentation.api.QualifiedClassNameCache;
 import io.opentelemetry.instrumentation.api.tracer.HttpClientTracer;
 import io.opentelemetry.trace.Span;
 import java.net.URI;
-import java.util.Map;
-import java.util.concurrent.ConcurrentHashMap;
 
 public class AwsSdkClientTracer extends HttpClientTracer<Request<?>, Request<?>, Response<?>> {
 
   static final String COMPONENT_NAME = "java-aws-sdk";
 
   public static final AwsSdkClientTracer TRACER = new AwsSdkClientTracer();
 
-  private final Map<String, String> serviceNames = new ConcurrentHashMap<>();
-  private final Map<Class, String> operationNames = new ConcurrentHashMap<>();
+  private final QualifiedClassNameCache cache =
```

Comment:
Ok so here's where it's used outside the package. I would stick to the old version, we could have it keyed on a struct of Class / String if it would help to bring down allocations. I don't know if I grokked the classnamecache correctly, but it seems to be a layer of fixedsize caches? Wouldn't expect it to win significantly to warrant this complexity.

Reply:
The motivation to use `ClassValue` is to handle class unloading properly.

Reply:
Can we stick with weak map if we need to handle class unload specially instead of the complex cache? Though since it's not static, not sure if there would be an unload issue but could be wrong.

Reply:
By unloading I’m referring to the classes themselves. ClassValue _is_ a weak map, just one optimised for class keys.

Reply:
Thanks for the context, glad to have learned a new tool :) I sent https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1153 which I think handles this case well with less complexity

Reply:
Fair enough, but the same pattern of constructing resource names with basic preprocessing and concatenation is all over the code base, and we'll be moving more of these over to use that class as they come up. IMHO the complexity you refer to comes from wanting to do in JDK7 what would have been simple with a JDK8 baseline.